go/ai/generate.go

```diff
@@ -86,8 +86,131 @@ func LookupModel(provider, name string) *Model {
+		req.Request.Messages = append(req.Request.Messages, NewSystemTextMessage(prompt))
+	}
+}
+
+// WithMessages adds provided messages to GenerateRequest.
+func WithMessages(messages ...*Message) GenerateParamsBuilder {
+	return func(req *GenerateParams) {
+		req.Request.Messages = append(req.Request.Messages, messages...)
+	}
+}
+
+// WithConfig adds provided config to GenerateRequest.
+func WithConfig(config any) GenerateParamsBuilder {
+	return func(req *GenerateParams) {
+		req.Request.Config = config
+	}
+}
+
+// WithCandidates adds provided candidate count to GenerateRequest.
+func WithCandidates(c int) GenerateParamsBuilder {
+	return func(req *GenerateParams) {
+		req.Request.Candidates = c
+	}
+}
+
+// WithContext adds provided context to GenerateRequest.
+func WithContext(c ...any) GenerateParamsBuilder {
+	return func(req *GenerateParams) {
+		req.Request.Context = c
+	}
+}
+
+// WithTools adds provided tools to GenerateRequest.
+func WithTools(tools ...Tool) GenerateParamsBuilder {
+	return func(req *GenerateParams) {
+		var toolDefs []*ToolDefinition
+		for _, t := range tools {
+			toolDefs = append(toolDefs, t.Definition())
+		}
+		req.Request.Tools = toolDefs
+	}
+}
+
+// WithOutputSchema adds provided output schema to GenerateRequest.
+func WithOutputSchema[T any](schema T) GenerateParamsBuilder {
+	return func(req *GenerateParams) {
+		if req.Request.Output == nil {
+			req.Request.Output = &GenerateRequestOutput{}
+			req.Request.Output.Format = OutputFormatJSON
+		}
+		req.Request.Output.Schema = base.SchemaAsMap(base.InferJSONSchemaNonReferencing(schema))
+	}
+}
+
+// WithOutputFormat adds provided output format to GenerateRequest.
+func WithOutputFormat(format OutputFormat) GenerateParamsBuilder {
+	return func(req *GenerateParams) {
+		req.Request.Output.Format = format
+	}
+}
+
+// WithStreaming adds a streaming callback to the generate request.
+func WithStreaming(cb ModelStreamingCallback) GenerateParamsBuilder {
+	return func(req *GenerateParams) {
+		req.Stream = cb
+	}
+}
+
+// Generate run generate request for this model. Returns GenerateResponse struct.
+func (m *Model) Generate(ctx context.Context, withs ...GenerateParamsBuilder) (*GenerateResponse, error) {
+	req := &GenerateParams{
+		Request: &GenerateRequest{},
+	}
+	for _, with := range withs {
+		with(req)
+	}
+
+	return m.GenerateRaw(ctx, req.Request, req.Stream)
+}
+
+// GenerateText run generate request for this model. Returns generated text only.
+func (m *Model) GenerateText(ctx context.Context, withs ...GenerateParamsBuilder) (string, error) {
+	res, err := m.Generate(ctx, withs...)
+	if err != nil {
+		return "", err
+	}
+
+	return res.Text(), nil
+}
+
+// Generate run generate request for this model. Returns GenerateResponse struct.
+// TODO: Stream GenerateData with partial JSON
+func (m *Model) GenerateData(ctx context.Context, value any, withs ...GenerateParamsBuilder) (*GenerateResponse, error) {
+	withs = append(withs, WithOutputSchema(value))
```

Comment:
This should probably throw an error if there are any other `WithOutputSchema`. At the very least, overwrite it but that's less deterministic behavior.

Reply:
added an error if WithOutputSchema  called more than once